chunky/src/java/se/llbit/math/Octree.java

```diff
@@ -490,19 +502,33 @@ public boolean enterBlock(Scene scene, Ray ray, BlockPalette palette) {
       ray.setCurrentMaterial(currentBlock);
 
       if (currentBlock.localIntersect) {
+        // Other functions expect the ray origin to be in the block they test so here time
+        // to update it
+        ray.o.scaleAdd(distance, ray.d);
+        ray.distance += distance;
+        distance = 0;
         if (currentBlock.intersect(ray, scene)) {
           if (prevBlock != currentBlock)
             return true;
 
           ray.o.scaleAdd(Ray.OFFSET, ray.d);
+          offsetX = -ray.o.x * invDx;
+          offsetY = -ray.o.y * invDy;
+          offsetZ = -ray.o.z * invDz;
           continue;
         } else {
           // Exit ray from this local block.
           ray.setCurrentMaterial(Air.INSTANCE); // Current material is air.
           ray.exitBlock(x, y, z);
+          offsetX = -ray.o.x * invDx;
+          offsetY = -ray.o.y * invDy;
+          offsetZ = -ray.o.z * invDz;
           continue;
         }
       } else if (!currentBlock.isSameMaterial(prevBlock) && currentBlock != Air.INSTANCE) {
+        // Origin and distance of ray need to be updated
+        ray.o.scaleAdd(distance, ray.d);
+        ray.distance += distance;
```

Comment:
`ray.o` and `ray.distance` are updated but `offsetX,Y,Z` and `invDx,y,z` are not?

Reply:
Yep because we are leaving the function so they are no longer needed